### PR TITLE
JAMES-1744 Preview should not contain HTML tags

### PR DESCRIPTION
**Pull request by Laura-Royet**

### CODE REVIEW

**Comment by aduprat** [comment]
That's a :+1: , but you may add more tests:
- some more complexe HTML
- some bad HTML (on opening tag without matching end tag, ...)

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java [hunk 1]
```diff
@@ -130,12 +132,13 @@ private static String getPreview(IndexableMessage im) {
...
-            return body;
+        String bodyWithoutTag = Jsoup.parse(body).text();
+        if (bodyWithoutTag.length() <= 256) {
+            return bodyWithoutTag;
         }
-        return body.substring(0, 253) + "...";
-    }
-    
+        return bodyWithoutTag.substring(0, 253) + "...";
+        }
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: indent

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 2]
```diff
@@ -214,7 +214,7 @@ public void getMessagesShouldReturnMessageWhenHtmlMessage() throws Exception {
             .body(ARGUMENTS + ".list[0].subject", equalTo("my test subject"))
             .body(ARGUMENTS + ".list[0].htmlBody", equalTo("This is a <b>HTML</b> mail"))
             .body(ARGUMENTS + ".list[0].isUnread", equalTo(true))
-            .body(ARGUMENTS + ".list[0].preview", equalTo("This is a <b>HTML</b> mail"))
+            .body(ARGUMENTS + ".list[0].preview", equalTo("This is a HTML mail"))
             .body(ARGUMENTS + ".list[0].headers", equalTo(ImmutableMap.of("content-type", "text/html", "subject", "my test subject")))
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: I would like to see a test were text is here

If the content type is text, HTML tag should not be removed IMO. Image I send a text mail like this : 

```
Content-type : text/plain

Little cousin, if you want to put some bold content, do it like this : <b>text in bold</b>
```

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java [hunk 3]
```diff
@@ -130,12 +132,13 @@ private static String getPreview(IndexableMessage im) {
     }
 
     @VisibleForTesting static String computePreview(String body) {
-        if (body.length() <= 256) {
-            return body;
+        String bodyWithoutTag = Jsoup.parse(body).text();
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: Compute preview might be done in an other class, so that preview is not coupled with message.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java [hunk 4]
```diff
@@ -130,12 +132,13 @@ private static String getPreview(IndexableMessage im) {
...
+        if (bodyWithoutTag.length() <= 256) {
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: Magic numbers

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java [hunk 5]
```diff
@@ -130,12 +132,13 @@ private static String getPreview(IndexableMessage im) {
...
+            return bodyWithoutTag;
         }
-        return body.substring(0, 253) + "...";
-    }
-    
+        return bodyWithoutTag.substring(0, 253) + "...";
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: Magic numbers again

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java (hunk 2 above)
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: Add a test over 256 characters please

Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: Already tested in MailboxMessageTest

Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: Ok for my second comment, not for the first :-)

File: server/protocols/jmap/pom.xml [hunk 6]
```diff
@@ -274,6 +274,21 @@
                     <artifactId>bcpkix-jdk15on</artifactId>
                 </dependency>
                 <dependency>
+                    <groupId>org.javatuples</groupId>
```
Comment (mbaechler) [inline server/protocols/jmap/pom.xml]: I understand you tried to fix the file ordering, but you failed at that because the current state is too bad.
Please make another commit to reorder everything properly.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java (hunk 3 above)
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: :+1:

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java (hunk 5 above)
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: and ellipsis is another feature, should be in another class too

Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: and I see no test about that

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java (hunk 3 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: Seems to be the same for lots of other static methods so what do think each member of James team? 
thanks

Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: it's just a strait SRP application, there's no much debate to have about that. Generating a preview should not be a Message responsibility.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreview.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
+package org.apache.james.jmap.model;
```
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreview.java]: license missing

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MailboxMessageTest.java [hunk 8]
```diff
@@ -353,7 +353,7 @@ public void bodyWith256LengthShouldNotBeTruncated() {
                 + "0000000000111111111122222222223333333333444444444455555555556666666666777777777788888888889999999999"
                 + "00000000001111111111222222222233333333334444444444555555";
         assertThat(body256.length()).isEqualTo(256);
-        assertThat(Message.computePreview(body256)).isEqualTo(body256);
+        assertThat(MessagePreview.getPreview(body256)).isEqualTo(body256);
```
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MailboxMessageTest.java]: Such test should be in a new class: MessagePreviewTest

**Comment by MichaelBailly** [comment]
Oh I like that PR ! No more HTML tags in unified inbox... Tx James team !

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreview.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+public class MessagePreview {
+    
+    public static final String NO_BODY = "(Empty)";
+    public static final int MAX_PREVIEW_LENGTH = 256;
+
+    public static String from(IndexableMessage message) {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreview.java]: you need some tests for that method

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 10]
```diff
@@ -220,6 +220,68 @@ public void getMessagesShouldReturnMessageWhenHtmlMessage() throws Exception {
     }
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: you still miss the "content-type=text/plain with some html tags in the body " test case

File: server/protocols/jmap/pom.xml (hunk 6 above)
Comment (mbaechler) [inline server/protocols/jmap/pom.xml]: you didn't handled this comment

File: server/protocols/jmap/pom.xml [hunk 11]
```diff
@@ -352,7 +347,7 @@
                     <artifactId>slf4j-simple</artifactId>
                     <scope>test</scope>
                 </dependency>
-            </dependencies>
+            </dependencies>
```
Comment (mbaechler) [inline server/protocols/jmap/pom.xml]: noise detected (:

Comment (Laura-Royet) [inline server/protocols/jmap/pom.xml]: ok. removed.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreview.java (hunk 7 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreview.java]: licence added.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java (hunk 10 above)
Comment (Laura-Royet) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: ok, treated.

File: server/protocols/jmap/pom.xml (hunk 6 above)
Comment (Laura-Royet) [inline server/protocols/jmap/pom.xml]: ok, done in another commit.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java (hunk 3 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: done.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java (hunk 4 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: ok, put in a constant.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java (hunk 5 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: both treated.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java (hunk 1 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: done

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreview.java (hunk 9 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessagePreview.java]: I added a class which test all the methods in this class.

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MailboxMessageTest.java (hunk 8 above)
Comment (Laura-Royet) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/MailboxMessageTest.java]: Ok, done (see previous comment).

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 12]
```diff
@@ -214,7 +214,69 @@ public void getMessagesShouldReturnMessageWhenHtmlMessage() throws Exception {
...
             .body(ARGUMENTS + ".list[0].htmlBody", equalTo("This is a <b>HTML</b> mail"))
             .body(ARGUMENTS + ".list[0].isUnread", equalTo(true))
-            .body(ARGUMENTS + ".list[0].preview", equalTo("This is a <b>HTML</b> mail"))
+            .body(ARGUMENTS + ".list[0].preview", equalTo("This is a HTML mail"))
+            .body(ARGUMENTS + ".list[0].headers", equalTo(ImmutableMap.of("content-type", "text/html", "subject", "my test subject")))
+            .body(ARGUMENTS + ".list[0].date", equalTo("2014-10-30T14:12:00Z"));
+    }
+    
+    @Test
+    public void getMessagesShouldReturnMessageWhenComplexeHtmlMessage() throws Exception {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: Complex without trailing e

Comment (Laura-Royet) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: ok, corrected.

**Comment by Laura-Royet** [comment]
Because of a problem linked to unstable test (now corrected), I have duplicated this branch and created a new PR : https://github.com/linagora/james-project/pull/266
